### PR TITLE
[ExportChiselInterface] Basic infrastructure for Chisel Interface file emission

### PR DESCRIPTION
**Pull request by trilorez**
This commit adds a pass for FIRRTL that generates a Scala file with a module
class that represents the interface for the top module of the FIRRTL circuit.
This is to support the development of separable compilation of FIRRTL circuits.
The details about linking circuits together are to be determined. The generated
Scala module extends ExtModule, although a new module class may be introduced
in the future.

### CODE REVIEW

File: include/circt/Conversion/Passes.td [hunk 1]
```diff
@@ -37,6 +37,40 @@ def AffineToPipeline : Pass<"convert-affine-to-pipeline", "mlir::func::FuncOp">
...
+    a FIRRTL circuit.
+  }];
+
+  let constructor = "createExportChiselInterfacePass()";
+  let dependentDialects = [
+    "circt::firrtl::FIRRTLDialect"
+  ];
+}
+
+def ExportSplitChiselInterface : Pass<"export-split-chisel-interface", "firrtl::CircuitOp"> {
```
Comment (trilorez) [inline include/circt/Conversion/Passes.td:57]: I went with the `Export/ExportSplit...` convention used for verilog and systemc emission even though only one file is exported for now.

File: lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+  os << "}\n";
+
+  return success();
+}
+
+/// Exports a Chisel interface to the output stream
+static LogicalResult exportChiselInterface(CircuitOp circuit,
+                                           llvm::raw_ostream &os) {
+  // Emit version, package, and import declarations
+  os << circt::getCirctVersionComment() << "package shelf."
```
Comment (trilorez) [inline lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp:181]: We'll probably want to make the package name configurable at some point but I wasn't sure if it was worth doing yet.

File: test/Conversion/ExportChiselInterface/basic.mlir [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      in %clock: !firrtl.clock,
+      in %uInt: !firrtl.uint<2>,
+      out %sInt: !firrtl.sint<4>,
+      out %analog: !firrtl.analog<5>,
+      in %bundle: !firrtl.bundle<a: uint<32>, b: analog<1>, c: bundle<d: sint<2>, e: sint<4>>>,
+      out %bundleWithFlip: !firrtl.bundle<word: uint<32>, valid: uint<1>, ready flip: uint<1>>,
+      in %vector: !firrtl.vector<uint<16>, 10>,
+      in %vectorOfVector: !firrtl.vector<vector<uint<16>, 10>, 20>,
+      in %vectorOfBundle: !firrtl.vector<bundle<word: uint<32>, valid: uint<1>, ready flip: uint<1>>, 5>) {}
+}
```
Comment (dtzSiFive) [inline test/Conversion/ExportChiselInterface/basic.mlir:47]: Micro-nit: newline at EOF.

File: tools/circt-opt/CMakeLists.txt [hunk 4]
```diff
@@ -1,9 +1,9 @@
 set(LLVM_LINK_COMPONENTS
   Support
-  )
+)
```
Comment (dtzSiFive) [inline tools/circt-opt/CMakeLists.txt:3]: Oof, looks like we're already inconsistent re:CMake formatting, thanks for touching it up.  This may be better if split into a separate PR.

File: tools/firtool/firtool.cpp [hunk 5]
```diff
@@ -228,6 +228,17 @@ static cl::opt<bool>
                 cl::desc("run the reset inference pass on firrtl"),
                 cl::init(true), cl::cat(mainCategory));
 
+static cl::opt<bool> exportChiselInterface(
```
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply tools/firtool/firtool.cpp:231]: This seems appropriate to me, although looks like this will output to stdout by default which is also the default output for firtool, causing both on stdout (not interleaved, but still).

File: tools/firtool/firtool.cpp [hunk 6]
```diff
@@ -228,6 +228,17 @@ static cl::opt<bool>
...
+    "export-chisel-interface",
+    cl::desc("generate a Scala Chisel interface to the top level "
+             "module of the firrtl circuit"),
+    cl::init(false), cl::cat(mainCategory));
+
+static cl::opt<std::string> chiselInterfaceOutDirectory(
```
Comment (dtzSiFive) [inline tools/firtool/firtool.cpp:237]: Directory being stdout ("-") seems a little strange, but I get this is a sentinel value of sorts to indicate it wasn't set and so should be emitted to stdout.

I'm guessing the directory option is not a filename so as to make it easier to dump interface files to a (re-used) location without needing to change this option to reflect a suitable filename?

File: tools/firtool/firtool.cpp [hunk 7]
```diff
@@ -536,6 +547,15 @@ processBuffer(MLIRContext &context, TimingScope &ts, llvm::SourceMgr &sourceMgr,
   if (inferResets)
     pm.nest<firrtl::CircuitOp>().addPass(firrtl::createInferResetsPass());
 
+  if (exportChiselInterface) {
+    if (chiselInterfaceOutDirectory == "-") {
```
Comment (dtzSiFive) [inline tools/firtool/firtool.cpp:551]: Nit: No braces around single-statement if/else blocks, so no braces needed around both of the `pm.nest...` lines.

File: tools/firtool/firtool.cpp [hunk 8]
```diff
@@ -904,6 +924,8 @@ int main(int argc, char **argv) {
     sv::registerPasses();
 
     // Export passes:
+    registerExportChiselInterfacePass();
```
Comment (dtzSiFive) [inline tools/firtool/firtool.cpp:927]: Thanks for keeping this maintained!

**Review by dtzSiFive — COMMENTED** [review]
This is neat, thank you! Some nits but mostly LGTM modulo some detalis that perhaps we iterate on and refine later.

File: include/circt/Conversion/ExportChiselInterface.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
+//===- ExportChiselInterface.h - Chisel Package Exporter ----------*- C++
+//-*-===//
```
[the remark this answers is not in the document]
Comment (trilorez) [reply include/circt/Conversion/ExportChiselInterface.h]: Oops, these were an issue with renaming, fixed

File: lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+#define DEBUG_TYPE "export-chisel-package"
+
+//===----------------------------------------------------------------------===//
+// Interface emission logic
+//===----------------------------------------------------------------------===//
+
+static const unsigned int indentIncrement = 2;
+
+/// Emits type construction expression for the port type, recursing into
+/// aggregate types as needed
```
[the remark this answers is not in the document]
Comment (trilorez) [reply lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp]: Updated

File: test/Conversion/ExportChiselInterface/basic.mlir (hunk 3 above)
Comment (trilorez) [reply test/Conversion/ExportChiselInterface/basic.mlir]: Added

File: tools/firtool/firtool.cpp (hunk 7 above)
Comment (trilorez) [reply tools/firtool/firtool.cpp:551]: If I remove any of the braces in these lines, I get the following warning:
> Add explicit braces to avoid dangling else

File: tools/firtool/firtool.cpp (hunk 6 above)
Comment (trilorez) [reply tools/firtool/firtool.cpp:237]: I agree `"-"` is a weird sentinel, I was using it since it is used elsewhere (see `inputFilename` and `outputFilename`). It's a directory since it makes sense to me implicitly name the generated file itself based on the circuit's top module name. But we can change this when we have a better idea of exactly what we want.

Comment (trilorez) [reply tools/firtool/firtool.cpp:237]: It looks like other file options use an empty string as a sentinel which makes more sense to me, so I've updated the option to use that

File: tools/firtool/firtool.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (trilorez) [reply tools/firtool/firtool.cpp:231]: I could remove the `exportChiselInterface` option and have `chiselInterfaceOutDirectory` only, maybe renamed to something like `chiselInterfaceExportDirectory`. If someone wanted to directly output a Chisel interface to stdout they could still use `circt-opt ... --export-chisel-interface`.

File: lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp (hunk 2 above)
Comment (darthscsi) [reply lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp:180]: Maybe default to the name of the module?

**Review by darthscsi — APPROVED** [review]
LGTM

File: lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp (hunk 2 above)
Comment (trilorez) [reply lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp:180]: Currently it's `shelf.<module-name>`
